yarn-project/prover-client/src/prover-pool/prover-pool.ts

```diff
@@ -0,0 +1,47 @@
+import { MemoryProvingQueue } from './memory-proving-queue.js';
+import { type ProvingAgent } from './prover-agent.js';
+import { type ProvingQueue } from './proving-queue.js';
+
+/**
+ * Utility class that spawns N prover agents all connected to the same queue
+ */
+export class ProverPool {
+  private agents: ProvingAgent[] = [];
+  private running = false;
+
+  constructor(
+    private size: number,
+    private agentFactory: (i: number) => ProvingAgent | Promise<ProvingAgent>,
+    public readonly queue: ProvingQueue = new MemoryProvingQueue(),
+  ) {}
+
+  async start(): Promise<void> {
+    if (this.running) {
+      throw new Error('Prover pool is already running');
+    }
+
+    // lock the pool state here since creating agents is async
+    this.running = true;
+
+    // protect against start, stop, start creating too many prover agents
+    if (this.agents.length < this.size) {
```

Comment:
This condition is redundant isn't it?

Reply:
Agent creation is deferred until `start` because the factory can return a Promise so we can enter this function with an empty array (the first time) or with a full array (after a start-stop cycle). We could create N new agents everytime instead and destroy them in `stop()`

Reply:
But doesn't the `if` statement just repeat the for loop's condition? Or am I being blind?

Reply:
No, I am bling, you're absolutely right!